security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,423 @@
+        },
+        connected: {
+          _onEnter: (callback) => {
+            callback(null);
+          },
+          getEndorsementKey: (callback) => {
+            callback(null, this._ek.asTpm2B());
+          },
+          getStorageRootKey: (callback) => {
+            callback(null, this._srk.asTpm2B());
+          },
+          signWithIdentity: (dataToSign, callback) => {
+            this._signData(dataToSign, (err: Error, signedData: Buffer) => {
+              if (err) {
+                debug('Error from signing data: ' + err);
+                this._fsm.transition('disconnected', callback, err);
+              } else {
+                callback(null, signedData);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._activateSymmetricIdentity(identityKey, (err: Error) => {
+              if (err) {
+                debug('Error from activate: ' + err);
+                this._fsm.transition('disconnected', callback, err);
+              } else {
+                callback(null);
+              }
+            });
+          },
+        }
+      }
+    });
+  }
+
+  getEndorsementKey(callback: (err: Error, endorsementKey: string) => void): void {
+      this._fsm.handle('getEndorsementKey', callback);
+  }
+
+  getStorageRootKey(callback: (err: Error, storageKey: string) => void): void {
+    this._fsm.handle('getStorageRootKey', callback);
+  }
+
+  signWithIdentity(dataToSign: Buffer, callback: (err: Error, signedData: Buffer) => void): void {
+    if (dataToSign === null || dataToSign.length === 0) {
+        throw new ReferenceError('\'dataToSign\' cannot be \'' + dataToSign + '\'');
+    }
+    if (this._idKeyPub == null) {
+        throw new errors.InvalidOperationError('activateSymmetricIdentity must be invoked before any signing is attempted.');
+    }
+    this._fsm.handle('signWithIdentity', dataToSign, callback);
+  }
+
+  activateSymmetricIdentity(identityKey: Buffer, callback: (err: Error, returnedActivate: Buffer) => void): void {
+    if (identityKey === null || identityKey.length === 0) {
+      throw new ReferenceError('\'identityKey\' cannot be \'' + identityKey + '\'');
+    }
+    this._fsm.handle('activateSymmetricIdentity', identityKey, callback);
+  }
+
+  getRegistrationId(callback: (err: Error, registrationId: string) => void): void {
+    if (this._registrationId) {
+      callback(null, this._registrationId);
+    } else {
+      this.getEndorsementKey( function (endorsementError: Error, endorsementKey: string): void {
+        if (endorsementError) {
+          callback(endorsementError, null);
+        } else {
+          let hasher = crypto.createHash('sha256');
+          hasher.update(endorsementKey);
+          this._registrationId = (base32Encode(hasher.digest(), 'RFC4648').toLowerCase()).replace(/=/g, '');
+          callback(null, this._registrationId);
+        }
+      }.bind(this));
```

Comment:
>.bind(this) [](start = 7, length = 11)

<soapbox> bind sucks!  you got something against self? </soapbox>
side question: doesn't typescript make bind and self unnecessary?  I want to say I saw a case where it was, but I think the debugger may have been lying to me.

Reply:
Agree.  My foolishness for making stuff that required it.  Thank you and the need for it has been removed.

---
In reply to: [159310972](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r159310972) [](ancestors = 159310972)